### ISSUE TITLE
Multiple sites failed to process during dmr pair

### ISSUE THREAD

**Issue by ArnavBharti** [nanoporetech/modkit#149]
Hello,

I was running `modkit dmr pair`. I got `finished, processed 695 sites successfully, 1760883 failed` output. What could be the reason so many sites failed to process?

OUTPUT:
```
> running single-site analysis
> using default prior, Beta(α: 0.55, β: 0.55)
> estimating max coverages from data
> sampled 891846 a records and 425260 b records, calculating max coverages for 95th percentile
> calculated max coverage for a: 6 and b: 4
> finished, processed 695 sites successfully, 1760883 failed
```

INPUT PREP:
```
modkit pileup all.bam pileup.bed
bgzip pileup.bed
tabix pileup.bed.gz
```

COMMAND RUN:
```
modkit dmr pair \
  -a sequencing/SEQUENCING_ANALYSIS/.../pass/pileup.bed.gz \
  -a sequencing/SEQUENCING_ANALYSIS/.../pass/barcode01/pileup.bed.gz \
  -a sequencing/SEQUENCING_ANALYSIS.../pass/barcode02/pileup.bed.gz \
  -a sequencing/SEQUENCING_ANALYSIS.../pass/barcode01/pileup.bed.gz \
  -a sequencing/SEQUENCING_ANALYSIS.../pass/barcode01/pileup.bed.gz \
  -b sequencing/SEQUENCING_ANALYSIS.../pass/barcode02/pileup.bed.gz \
  -b sequencing/SEQUENCING_ANALYSIS.../pass/barcode10/pileup.bed.gz \
  --ref PlasmoDB-61_....fasta \
  --base C \
  -o dmr_ed_yr_eb_fy_el_ej_67 \
  -t 10 \
  -f
```

**Comment by ArtRand** [comment]
Hello @ArnavBharti, 

I'm assuming you're using 0.2.6, correct (the newest version has the fix from the build I gave you before). Could you run the command with `--log-filepath` and send me the file it produces?

**Comment by ArnavBharti** [comment]
Yes, I am using 0.2.6.

LOGS:
```
[src/logging.rs::52][2024-03-18 21:49:37][DEBUG] command line: modkit dmr pair -a sequencing/SEQUENCING_ANALYSIS/5mc_FY_basecalling/pass/pileup.bed.gz -a sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode01/pileup.bed.gz -a sequencing/SEQUENCING_ANALYSIS_1/YR_CC_5MC_BASECALLING/pass/barcode02/pileup.bed.gz -a sequencing/SEQUENCING_ANALYSIS_1/YQ_YR_5mC_Basecalling/pass/barcode01/pileup.bed.gz -a sequencing/SEQUENCING_ANALYSIS/EB_EL_5mC/pass/barcode01/pileup.bed.gz -b sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode02/pileup.bed.gz -b sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode03/pileup.bed.gz -b sequencing/SEQUENCING_ANALYSIS/5mc_basecalling_et_ev_ej/pass/barcode11/pileup.bed.gz -b sequencing/SEQUENCING_ANALYSIS/5mc_basecalling_et_ev_ej/pass/barcode12/pileup.bed.gz --ref PlasmoDB-61_PvivaxSal1_Genome.fasta --base C -o dmr_ed_yr_eb_fy_cc_cq_et_ev -t 10 -f --log-filepath ./dmr_log
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/5mc_FY_basecalling/pass/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/5mc_FY_basecalling/pass/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode01/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode01/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS_1/YR_CC_5MC_BASECALLING/pass/barcode02/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS_1/YR_CC_5MC_BASECALLING/pass/barcode02/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS_1/YQ_YR_5mC_Basecalling/pass/barcode01/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS_1/YQ_YR_5mC_Basecalling/pass/barcode01/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/EB_EL_5mC/pass/barcode01/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/EB_EL_5mC/pass/barcode01/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode02/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode02/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode03/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/ed_cc_cq_5mc_basecalling/pass/barcode03/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/5mc_basecalling_et_ev_ej/pass/barcode11/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/5mc_basecalling_et_ev_ej/pass/barcode11/pileup.bed.gz.tbi
[src/dmr/tabix.rs::81][2024-03-18 21:49:37][DEBUG] looking for index associated with sequencing/SEQUENCING_ANALYSIS/5mc_basecalling_et_ev_ej/pass/barcode12/pileup.bed.gz at sequencing/SEQUENCING_ANALYSIS/5mc_basecalling_et_ev_ej/pass/barcode12/pileup.bed.gz.tbi
[src/dmr/subcommands.rs::281][2024-03-18 21:49:37][INFO] reading reference FASTA at "PlasmoDB-61_PvivaxSal1_Genome.fasta"
[src/dmr/subcommands.rs::296][2024-03-18 21:49:37][INFO] running single-site analysis
[src/dmr/single_site.rs::67][2024-03-18 21:49:37][INFO] using default prior, Beta(α: 0.55, β: 0.55)
[src/dmr/single_site.rs::675][2024-03-18 21:49:37][INFO] estimating max coverages from data
[src/dmr/single_site.rs::719][2024-03-18 21:49:37][INFO] sampled 891846 a records and 425260 b records, calculating max coverages for 95th percentile
[src/dmr/single_site.rs::744][2024-03-18 21:49:37][INFO] calculated max coverage for a: 6 and b: 4
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
[src/dmr/single_site.rs::217][2024-03-18 21:49:39][DEBUG] score error, don't have counts from all samples
```

**Comment by ArtRand** [comment]
Hello @ArnavBharti,

It appears that you have a lot of errors that look like this:

> score error, don't have counts from all samples

You'll see this error when modkit tries to calculate the differential methylation metrics at a genomic position but cannot find bedMethyl records (i.e. counts of how many reads called each modification type) for _every_ sample you provided. For example, looks like you have 5 'a' samples and 2 'b' samples, say at a genomic position only 4 of the 'a' samples have a bedMethyl record, you'll see this error. I put this check in because I didn't want some of your DMR records to report on all of the samples and some to report on a subset of the samples. 

From the log line:
> calculated max coverage for a: 6 and b: 4

it looks like the coverage you have is on the lower side, so it might not be completely unreasonable that there are a number of positions with zero coverage in at least one sample. One thing you _may_ try, but it's a reach, if you're interested in CpG methylation or some other reverse-complement symmetric methylation you could run `modkit pileup` with the `--combine-strands --motif CG 0` options, (replace `--motif CG 0` with whatever motif you are looking for). You'll combine calls from the (+) and (-) strands, so you'll get a higher coverage numbers. But if your runs simply don't have any reads in common regions, there's no way around that.

**Comment by billytcl** [comment]
We are having this exact same issue. Is there a way to "relax" the dmr pipeline so that it won't fail? Things like t-test in R are able to deal with missing values, and it would be nice to use your pipeline instead. In my mind for each CpG site it would just first do a missing value filter first (eg. filter out all the samples that can't grab a site using tabix index), and then do the statistical test? Eg. we would like to be able to consider and test CpG sites that have >2 samples for each group.

I guess the dumb workaround would be to do an aggregation of all the "controls" and "cases" and then run modkit, but then we lose out on the N for each group....

**Comment by billytcl** [comment]
Actually, does dmr pair in this case consider the N samples in each group in its calculations? When I read the documentation it doesn't seem like that's the case but maybe I'm incorrect here. Because if it doesn't consider N then we should just aggregate the calls by merging the bams and then using modkit.

**Comment by ArtRand** [comment]
Hello @billytcl and @ArnavBharti,

Let me answer the second comment first. 

> Actually, does dmr pair in this case consider the N samples in each group in its calculations? When I read the documentation it doesn't seem like that's the case but maybe I'm incorrect here. Because if it doesn't consider N then we should just aggregate the calls by merging the bams and then using modkit.

When multiple samples are provided `dmr pair` simply aggregates the calls within each condition. There is also a "balanced" MAP-based p-value and effect size [see column 16 and 17](https://nanoporetech.github.io/modkit/intro_dmr.html#differential-methylation-output-format) where the counts from each sample are re-weighted so they have equivalent impact on the metric. Currently, you'll only see this metric when you have an equivalent number of `a` samples and `b` samples, but that's a bug and I'll get a fix out asap. 

> Is there a way to "relax" the dmr pipeline so that it won't fail? Things like t-test in R are able to deal with missing values, and it would be nice to use your pipeline instead. 

I can remove the constraint, it's not actually a requirement for the MAP-based p-value or the likelihood-ratio score. My main concern was that (in some experiments I looked at) it's possible that one sample within a condition has a different distribution of methylation at a location than the others of that condition. So to use an extreme example, given 3 'a' samples, say the fraction of 5mC was (0.9, 0.95, 0.1). If the last one (10% 5mC) was missing from the calculation at a specific site, you would potentially see a larger effect size than expected. I can remove the constraint and add a column with the % of samples contained in the statistical test, then you can decide to filter them if you like. Before you ask, I'd rather not go down the imputation route. What do you think?

**Comment by ArtRand** [comment]
@ArnavBharti @billytcl 

Give this a try. 

[modkit_dev439cfbed_centos7_x86_64.tar.gz](https://github.com/nanoporetech/modkit/files/14643412/modkit_dev439cfbed_centos7_x86_64.tar.gz)

**Comment by billytcl** [comment]
Thank you! We will try this out. And yes, we'd rather not do imputation!

**Comment by ArnavBharti** [comment]
Thank you!

What are the column headers in `dmr pair` for this build? There are 19 columns being created.

> @ArnavBharti @billytcl
> 
> Give this a try.
> 
> [modkit_dev439cfbed_centos7_x86_64.tar.gz](https://github.com/nanoporetech/modkit/files/14643412/modkit_dev439cfbed_centos7_x86_64.tar.gz)

**Comment by ArtRand** [comment]
@ArnavBharti 

The two new columns have the % of 'a' samples used and % of 'b' samples used in the test ([code](https://github.com/nanoporetech/modkit/blob/439cfbed16282a93559e462e2a28dc8a8d65e1bf/src/dmr/single_site.rs#L561)). You can also run with `--header` and you'll get a header line.

**Comment by ArtRand** [comment]
This fix is incorperated into [v0.2.7](https://github.com/nanoporetech/modkit/releases/tag/v0.2.7), please re-open this issue if you encounter any problems.
